scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -765,23 +748,38 @@ def get_matches_commandline(self, match_pattern):
 
     def get_child_processes(self, ppid):
         """
-        Given a process pid, get the children pids
-        @param ppid: parent process id
-        @return: list of children process IDs
+        Given a process id, return all children processes (recursively)
+        :param ppid: parent process id
+        :return: list of all children process ids
         """
 
         all_children = []
-        children_to_explore = []
+        children_to_explore = set()
         for _pid in self.parent_to_children_map[ppid]:
             all_children.append(_pid)
-            children_to_explore.append(_pid)
+            children_to_explore.add(_pid)
 
         # get the children 'recursively'
         while children_to_explore:  # the invariant
             child_to_explore = children_to_explore.pop()
-            all_children.extend(self.parent_to_children_map[child_to_explore])
+            if not self.parent_to_children_map.get(child_to_explore):
+                continue
+            unvisited = self.parent_to_children_map[child_to_explore]
+            for node in unvisited:
+                children_to_explore.add(node)
+                all_children.append(node)
```

Comment:
Just for safety, check first to see if `node` is in `all_children` and only add it to `children_to_explore` if it is not.

Reply:
I can do that, but for the same reason of duplicity,  we're removing duplicates 2 lines later `list(set(all_children))` but it doesn't hurt checking.